From 0dd6e91340911ee7a90987d4b2bf0ddf9768585c Mon Sep 17 00:00:00 2001
From: Shreenidhi Shedi <sshedi@vmware.com>
Date: Mon, 11 Jul 2022 11:54:59 +0000
Subject: [PATCH] gh-95231: Handle the EPERM error gracefully in crypt.py.

If kernel fips is enabled, we get permission error upon doing
`import crypt`.

Logs after reproducing the issue:
Python 3.9.1 (default, May 10 2022, 11:36:26)
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import crypt
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.9/crypt.py", line 117, in <module>
    _add_method('MD5', '1', 8, 34)
  File "/usr/lib/python3.9/crypt.py", line 94, in _add_method
    result = crypt('', salt)
  File "/usr/lib/python3.9/crypt.py", line 82, in crypt
    return _crypt.crypt(word, salt)
PermissionError: [Errno 1] Operation not permitted

Signed-off-by: Shreenidhi Shedi <sshedi@vmware.com>
---
 Lib/crypt.py                                                   | 3 +++
 .../next/Library/2022-07-25-15-45-06.gh-issue-95231.i807-g.rst | 3 +++
 2 files changed, 6 insertions(+)
 create mode 100644 Misc/NEWS.d/next/Library/2022-07-25-15-45-06.gh-issue-95231.i807-g.rst

diff --git a/Lib/crypt.py b/Lib/crypt.py
index 46c3de8474..92e70415e1 100644
--- a/Lib/crypt.py
+++ b/Lib/crypt.py
@@ -100,6 +100,9 @@ def _add_method(name, *args, rounds=None):
         # Not all libc libraries support all encryption methods.
         if e.errno == errno.EINVAL:
             return False
+        # unsupported or blocked by crypto policy
+        if e.errno in {errno.EPERM, errno.ENOSYS}:
+            return False
         raise
     if result and len(result) == method.total_size:
         methods.append(method)
diff --git a/Misc/NEWS.d/next/Library/2022-07-25-15-45-06.gh-issue-95231.i807-g.rst b/Misc/NEWS.d/next/Library/2022-07-25-15-45-06.gh-issue-95231.i807-g.rst
new file mode 100644
index 0000000000..33d7fd7c58
--- /dev/null
+++ b/Misc/NEWS.d/next/Library/2022-07-25-15-45-06.gh-issue-95231.i807-g.rst
@@ -0,0 +1,3 @@
+Handle EPERM errors gracefully in Lib/crypt.py
+One of the scenarios we hit this condition is when kernel fips is enabled
+and trying to use md5 from crypt module.
-- 
2.30.3

